analytics/input_redactor.go

```diff
@@ -0,0 +1,274 @@
+			return append(container, value.Interface()), append(expandedContainer, value.Interface()), nil
+		} else {
+			return append(container, ""), append(expandedContainer, ""), nil
+		}
+	}
+}
+
+func (i inputRedactor) fillMap(options models.EnvironmentItemOptionsModel, container map[string]interface{}, expandedContainer map[string]interface{}, key string, value reflect.Value) error {
+	for value.Kind() == reflect.Ptr || value.Kind() == reflect.Interface {
+		value = value.Elem()
+	}
+	switch value.Kind() {
+	case reflect.Array, reflect.Slice:
+		acc, expandedAcc, err := i.collectSliceItems(options, value)
+		if err != nil {
+			return err
+		}
+		container[key] = acc
+		expandedContainer[key] = expandedAcc
+	case reflect.Map:
+		acc, expandedAcc, err := i.collectMapItems(options, value)
+		if err != nil {
+			return err
+		}
+		container[key] = acc
+		expandedContainer[key] = expandedAcc
+	case reflect.String:
+		v, e, err := i.expandAndRedactString(options, value)
+		if err != nil {
+			return err
+		}
+		container[key] = v
+		expandedContainer[key] = e
+	default:
+		if value.IsValid() && value.CanInterface() {
+			container[key] = value.Interface()
+			expandedContainer[key] = value.Interface()
+		} else {
+			container[key] = ""
+			expandedContainer[key] = ""
+		}
+	}
+	return nil
+}
+
+func (i inputRedactor) collectMapItems(options models.EnvironmentItemOptionsModel, value reflect.Value) (map[string]interface{}, map[string]interface{}, error) {
+	acc := map[string]interface{}{}
+	expandedAcc := map[string]interface{}{}
+	iter := value.MapRange()
+	for iter.Next() {
+		k := iter.Key()
+		v := iter.Value()
+		if err := i.fillMap(options, acc, expandedAcc, fmt.Sprintf("%s", k), v); err != nil {
+			return nil, nil, err
+		}
+	}
+	return acc, expandedAcc, nil
+}
+
+func (i inputRedactor) collectSliceItems(options models.EnvironmentItemOptionsModel, value reflect.Value) ([]interface{}, []interface{}, error) {
+	var acc []interface{}
+	var expandedAcc []interface{}
+	var err error
+	for j := 0; j < value.Len(); j++ {
+		acc, expandedAcc, err = i.growSlice(options, acc, expandedAcc, value.Index(j))
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	return acc, expandedAcc, nil
+}
+
+func (i inputRedactor) expandAndRedactString(options models.EnvironmentItemOptionsModel, value reflect.Value) (string, string, error) {
+	v := value.String()
+	e, err := i.expand(options, v)
+	if err != nil {
+		return "", "", err
+	}
+	v, err = RedactStringWithSecret(v, i.secrets)
+	if err != nil {
+		return "", "", err
+	}
+	e, err = RedactStringWithSecret(e, i.secrets)
+	if err != nil {
+		return "", "", err
+	}
+	return v, e, nil
+}
+
+func (i inputRedactor) expand(options models.EnvironmentItemOptionsModel, value string) (string, error) {
+	result := value
+	var envs map[string]string
+	if options.IsTemplate != nil && *options.IsTemplate {
+		if envs == nil {
+			effects, err := env.GetDeclarationsSideEffects(i.environment, i.envSource)
```

Comment:
This is a mixture of reusing and reimplementing env expansion and secret redaction found in the code. Without huge refactoring effort this was the best I could do to isolate duplicated logic in this file.

Reply:
I'll check out what I can do to reduce code duplication